tools/codegen/gen.py

```diff
@@ -327,7 +337,118 @@ def func(f: NativeFunction) -> Optional[str]:
+            # Only TypeDefault generation we do is registration
+            if dispatch is None and target is not Target.REGISTRATION:
+                return []
+
+            functional_func = g.out.func.signature()
+            functional_sig = DispatcherSignature.from_schema(functional_func)
+            meta_name = meta.name(functional_func)
+
+            # This is a little abusive; this assumes that the functionalization
+            # transformation ALWAYS refers to valid arguments in the original
+            # signature
+            functional_exprs = ', '.join(e.expr for e in functional_sig.exprs())
+
+        @with_native_function
+        def subfunc(f: NativeFunction) -> str:
+            k = f.func.kind()
+            sig = NativeSignature.from_schema(f.func)
+
+            if target is Target.DECLARATION:
+                # Not necessary!  Only TypeDefault is actually used, see
+                # https://github.com/pytorch/pytorch/issues/46319
+                return ""
+            elif target is Target.DEFINITION:
+                assert dispatch is not None
+                assert g.out is not None  # sigh
+                assert g.out.dispatch is not None
+
+                out_impl_name = f"at::native::{g.out.dispatch[dispatch]}"
+
+                # TODO: work a little harder to generate fresh names here
+                # TODO: less praying that I picked the right argument names
+
+                if k is SchemaKind.functional:
+                    out_expr = "result"
+                    prologue = "auto result = tensor_from_meta(meta_result);"
+                elif k is SchemaKind.inplace:
+                    out_expr = "self"
+                    prologue = "// TODO: consistency check assert"
+                elif k is SchemaKind.out:
+                    # TODO: generalize this for multi-out
+                    assert len(f.func.out_arguments) == 1, "multi-out structured not supported yet"
+                    # TODO: naughtily leaning on some codegen invariants....
+                    out_expr = f.func.out_arguments[0].name
+                    prologue = f"""
+// TODO: add a consistency check for meta_result
+{out_expr}.resize_(meta_result.sizes);
+"""
+
+                device_guard = ""
+
+                if dispatch is None or 'CUDA' in dispatch or 'Vulkan' == dispatch:
+                    # TODO: stop copypaste
+                    self_args = (a for a in f.func.arguments if a.name == "self")
+                    candidate_args = itertools.chain(self_args, f.func.out_arguments, f.func.arguments)
+                    device_of = next((f'{a.name}' for a in candidate_args if a.type.is_tensor_like()), None)
+
+                    device_guard = ''
+                    if f.device_guard and device_of is not None:
+                        # TODO: OptionalCUDAGuard
+                        device_guard = f"""\
+    const OptionalDeviceGuard device_guard(device_of({device_of}));
+"""
+
+                return f"""\
```

Comment:
When I studied the codegen, templates like this helps me learn the connection between internal functions and the final output; but sometimes I felt seeing a little concrete output example could be more intuitive (the downside is that we need remember to update the example). Maybe including a link to the RFC above structured_func()?